concurrency/MpscMemoryQueue.h

```diff
@@ -0,0 +1,155 @@
+#pragma once
+#include "utils/Allocator.h"
+#include <algorithm>
+#include <atomic>
+#include <cassert>
+#include <memory>
+
+namespace concurrency
+{
+/**
+ * Multiple producer single consumer queue of arbitrary large memory blocks.
+ * Useful as wait free queue for variable size items in mpsc setting.
+ */
+class MpscMemoryQueue
+{
+    enum CellState : uint32_t
+    {
+        emptySlot = 0,
+        allocated,
+        committed,
+        padding
+    };
+
+    struct Guard
+    {
+        Guard() {}
+        explicit Guard(bool allocated)
+        {
+            if (allocated)
+            {
+                std::memcpy(pattern, "ABXYGAHAERLKBOSP", 16);
+            }
+        }
+
+        bool operator==(const Guard& o) const { return 0 == std::memcmp(pattern, o.pattern, sizeof(pattern)); }
+
+        char pattern[16] = {0};
+    };
+
+    struct Entry
+    {
+        Entry() : state(CellState::emptySlot) {}
+        size_t entrySize() const { return entryOverHead() + size; }
+        static constexpr uint32_t headSize() { return sizeof(Entry); }
+        static constexpr uint32_t entryOverHead()
+        {
+#ifdef DEBUG
+            return sizeof(Entry) + sizeof(Guard);
+#else
+            return sizeof(Entry);
+#endif
+        }
+
+        static Entry* fromPtr(void* p)
+        {
+            return reinterpret_cast<Entry*>(reinterpret_cast<uint8_t*>(p) - sizeof(Entry));
+        }
+
+        void checkGuards() const;
+        Guard& tailGuard() { return *reinterpret_cast<Guard*>(data + size); }
+        const Guard& tailGuard() const { return *reinterpret_cast<const Guard*>(data + size); }
+#ifdef DEBUG
+        Guard frontGuard;
+#endif
+        std::atomic<CellState> state;
+        uint32_t size;
+        uint8_t data[];
```

Comment:
This only works for structures with alignment <= 32 bits

Reply:
will fix